server/server.go

```diff
@@ -177,27 +187,30 @@ func (h *handler) handle(pkt *protocol.Packet) {
 	}
 	h.tokens <- struct{}{}
 	h.mtx.Lock()
-	_, err := respPkt.WriteTo(h.conn)
-	h.mtx.Unlock()
+	defer h.mtx.Unlock()
+	err := h.conn.SetWriteDeadline(time.Now().Add(h.timeout))
 	if err != nil {
-		log.Errorf("connection %v: error in writing response %v", h.name, err)
-		h.conn.Close() // ignoring error: what can we do?
-		h.s.mtx.Lock()
-		delete(h.s.listeners[h.listener], h.conn)
-		h.s.mtx.Unlock()
-		logConnFailure()
+		h.closeWithWritingErr(err)
+	}
+	_, err = respPkt.WriteTo(h.conn)
```

Comment:
You shouldn't execute this if you closed the connection due to `SetWriteDeadline` returning an error